File: map-service/src/main/java/com/example/map/service/CircleService.java
package com.example.map.service;

import com.example.map.dto.CircleCreateRequest;
import com.example.map.dto.CircleResponse;
import com.example.map.entity.Circle;
import com.example.map.entity.CircleMember;
import com.example.map.repository.CircleMemberRepository;
import com.example.map.repository.CircleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Stream;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class CircleService {

    private final CircleRepository circleRepository;
    private final CircleMemberRepository circleMemberRepository;

    @Transactional
    public CircleResponse createCircle(CircleCreateRequest request, Long userId) {
        Circle circle = Circle.create(request.getName(), userId);
        Circle saved = circleRepository.save(circle);

        CircleMember ownerMember = CircleMember.create(saved, userId);
        circleMemberRepository.save(ownerMember);

        return CircleResponse.from(saved);
    }

    public List<CircleResponse> getMyCircles(Long userId) {
        List<Circle> ownedCircles = circleRepository.findByOwnerId(userId);

        List<Circle> memberCircles = circleMemberRepository.findByUserId(userId).stream()
                .map(CircleMember::getCircle)
                .toList();

        return Stream.concat(ownedCircles.stream(), memberCircles.stream())
                .distinct()
                .map(CircleResponse::from)
                .toList();
    }
}

File: apps/map-service/src/circle/circle.service.ts
import { Injectable } from '@nestjs/common';
import { CircleRepository } from './circle.repository';
import { CircleMemberRepository } from './circle-member.repository';
import { CreateCircleDto } from './dto/create-circle.dto';
import { CircleResponseDto } from './dto/circle-response.dto';
import { Circle } from './circle.entity';
import { CircleMember } from './circle-member.entity';

@Injectable()
export class CircleService {
  constructor(
    private readonly circleRepository: CircleRepository,
    private readonly circleMemberRepository: CircleMemberRepository,
  ) {}

  async createCircle(dto: CreateCircleDto, userId: number): Promise<CircleResponseDto> {
    const circle = new Circle();
    circle.name = dto.name;
    circle.ownerId = userId;
    circle.members = [];

    const saved = await this.circleRepository.save(circle);

    const ownerMember = new CircleMember();
    ownerMember.circle = saved;
    ownerMember.userId = userId;
    await this.circleMemberRepository.save(ownerMember);

    saved.members = [ownerMember];
    return CircleResponseDto.from(saved);
  }

  async getMyCircles(userId: number): Promise<CircleResponseDto[]> {
    const ownedCircles = await this.circleRepository.findByOwnerId(userId);
    const memberships = await this.circleMemberRepository.findByUserId(userId);
    const memberCircles = memberships.map((m) => m.circle);

    const uniqueCircles = new Map<number, Circle>();
    [...ownedCircles, ...memberCircles].forEach((c) => uniqueCircles.set(c.id, c));

    return Array.from(uniqueCircles.values()).map(CircleResponseDto.from);
  }
}
